test/player.spec.js

```diff
@@ -28,4 +28,31 @@ describe("Player's logic", () => {
   it('When player answered and onQuestionAnswered function is undefined, check if there is no error', () => {
     expect(() => player.answer(chosenAnswer)).not.toThrow();
   });
+
+  it('When computer player is asked, check if question in computer mind is question', () => {
+    const question = {
+      answers: ['Luke Skywalker', 'R2-D2', 'Chewbacca', 'Boba Fett'],
+      image: 'c3RhdGljL2Fzc2V0cy9pbWcvbW9kZXMvcGVvcGxlLzEuanBn',
+      rightAnswer: 'Luke Skywalker',
+    };
+    const computerPlayer = new Player();
+    computerPlayer.askQuestion(question, computerMind.setQuestion);
+    expect(computerMind.question).toBe(question);
+  });
+
+  it('When computer player answered, check if computer mind called function', () => {
```

Comment:
Doceniam bardzo Twoje zaangażowanie, ale to Piotrek niestety nie działa :/ 
Twój test nie sprawdza jak działa ten computerMind.  Mogę sobie zrobić tak:
```
 it('When computer player answered, check if computer mind called function', () => {
    const question = {
      answers: ['Luke Skywalker', 'R2-D2', 'Chewbacca', 'Boba Fett'],
      image: 'c3RhdGljL2Fzc2V0cy9pbWcvbW9kZXMvcGVvcGxlLzEuanBn',
      rightAnswer: 'Luke Skywalker',
    };
    const computerPlayer = new Player();
    computerPlayer.askQuestion(question, computerMind.setQuestion);
    let isFunctionCalled = false;
    computerPlayer.answer(
      {},
      () => (isFunctionCalled = true),
    );
    expect(isFunctionCalled).toBe(true);
  });
  ```
  I jest zielono. Test dalej testuje tylko działanie, że computerPlayer.answer() wywołuje callback.
  Tego computerMind.tryToAnswer też nigdzie nie wywoałałeś - tutaj przekazałeś referencje do funkcji.
  Jak piszesz testy najpierw trzeba zrobić RED (masz być pewien, że test nie przechodzi), potem GREEN (ma przejsc dopiero po implementacji, zobacz sobie czy jak cokolwiek zmienisz w implementacji to jest RED, wtedy znaczy, ze test jest dobry), dopiero wtedy robisz REFACTOR (poprawiasz zmienne, wyciagasz funkcje).
  
  Sam intellij by się przydał nawet bo już podkreśla sporo rzeczy:
  
  
![image](https://user-images.githubusercontent.com/18637116/104121248-18da4400-533d-11eb-9653-c27938619675.png)

![image](https://user-images.githubusercontent.com/18637116/104121251-1ed02500-533d-11eb-8d8c-1f03b8dac469.png)

![image](https://user-images.githubusercontent.com/18637116/104121255-22fc4280-533d-11eb-825a-44ef95afbcc1.png)

Nawet są literówki (length - o tym Cię nie poinformuje nawet wywołanie kodu bo setTimeout jest async - więc ono nie zdaży się wykonać) w zmiennych czy złe typy. Jednak IDE pomaga takie problemy zobaczyć od razu i niestety potwierdza to co zdawało się moim oczom wcześniej. ALE dobrze napisane testy są o wiele więcej warte i żadne IDE nie jest potrzebne :) 
Jak sprawdzisz sobie coverage to widać, że te linijki nie są w ogóle pokryte:
![image](https://user-images.githubusercontent.com/18637116/104123049-17167d80-5349-11eb-8ad1-e3bee4747488.png)

Może to spróbuje Ci pomóc (ja bym tak to widział). Ten nasz ComputerMind ma działać trochę jak nasz mózg - więc reagować na bodźce i móc coś zrobić.

Czyli. Teraz tak, początek testu jest dobry.
```
const computerPlayer = new Player();
computerPlayer.askQuestion(question, computerMind.setQuestion);
```
Robimy gracza i zadajemy mu pytanie. OK
To co bym zmienił to od razu mówimy komputerobi - no niech zareaguje, czyli.
`computerPlayer.askQuestion(question, computerMind.tryToAnswer);`
Dostaliśmy pytanie, to się STAŁO, więc callback - próbujmy odpowiedzieć. 
Wtedy nie potrzebujemy przetrzymywać tego question.
To też rodzi problem co jakby ktoś wywołał setQuestion ponownie. Lepiej unikać takiego stanu w kodzie :) 

Wtedy jak już mamy pytanie, to computer mind potrzebuje referencji do gracza, bo będzie musiał kazać mu odpowiedzieć. I to jest nasza asercja, którą powinien sprawdzać test. Czyli czy jeśli my zrobimy:
```
computerPlayer.askQuestion(question, computerMind.tryToAnswer);
expect(computerPlayer.answer).wasCalledWith(iTutajJakasOdpowiedzKtoraWygenerowalTryAnswer)
```
Czyli tryToAnswer - by nam odpowiadało za coś w tym stylu:
1. O! Mam pytanie, spróbuję odpowiedzieć
2. Myślę (mija czas - funkcja async!), po ilus tam sekundach znam odpowiedz (dobrze juz to zasymulowac, bo np. request do serwera o rozpoznaie obrazu tez zajmuje czas) - tak jak zrobiles (prawie).
3. Mam juz jakas odpowiedz - udzielam jej. Na player wywoluje player.answer(toCoUstalilemJakoOdpowiedz).
Coś tak w tym stylu :) Tak to przynajmniej widzę. 

I chyba tryToAnswer wtedy bedzie musial miec jeszcze jeden parametr onAnswered jako callback, który przekaże do compuerPlayer.answer(odpowiedzKtoraUstalil, onAnswered).

Dłuuugie, nie wiem czy zrozumiałeś :) PRAWIE dobrze kombinujesz, ale dobrze napisany test na początku dałby Ci pewność, że to teraz niestety nie śmiga.

Reply:
Ok, za 42min siadam do komputera i to ogarnę.
Jestem jeszcze zielony w testach.
(Po wczoraj już widzę ich niezbędność, a po dziś, że trzeba to jeszcze robić dobrze)

Reply:
Proponuję, że możemy test napisać razem jak coś :) 
To wtedy już z implementacja polecisz.
Nawet niegłupim pomysłem na przyszłość byloby - pierwszy PR z samym testem (do akceptacji) jak test zaakceptowany to lecimy z implementacja. 

Mam nadzieję, że się nie zrazisz :) Niestety muszę dawać te uwagi - ale w prawdziwych projektach będą na wagę złota. Tutaj jesteśmy po to, żeby się uczyć.